File: archive/pulumi-py/Pr7609/lib/secrets_stack.py
"""
secrets_stack.py

AWS Secrets Manager for database credentials.
Securely stores RDS PostgreSQL credentials with encryption.
"""

from typing import Optional, Dict
import json
import pulumi
from pulumi import ResourceOptions
from pulumi_aws import secretsmanager
import pulumi_random as random


class SecretsStack(pulumi.ComponentResource):
    """
    Secrets Manager stack for database credentials.

    Creates:
    - Random password for RDS
    - Secrets Manager secret with database credentials
    - Encryption at rest
    """

    def __init__(
        self,
        name: str,
        environment_suffix: str,
        tags: Optional[Dict] = None,
        opts: Optional[ResourceOptions] = None
    ):
        super().__init__('tap:secrets:SecretsStack', name, None, opts)

        resource_tags = tags or {}

        # Generate random password for database
        self.db_password = random.RandomPassword(
            f"db-password-{environment_suffix}",
            length=32,
            special=True,
            override_special="!#$%&*()-_=+[]{}<>:?",
            opts=ResourceOptions(parent=self)
        )

        # Create database username
        db_username = "japancart_admin"

        # Create secret with credentials
        # Build secret value using Output.concat to avoid json.dumps in apply()
        secret_value = pulumi.Output.concat(
            '{"username": "', db_username,
            '", "password": "', self.db_password.result,
            '", "engine": "postgres", "host": "placeholder", "port": 5432, "dbname": "transactions"}'
        )

        self.db_secret = secretsmanager.Secret(
            f"db-credentials-{environment_suffix}",
            name=f"db-credentials-{environment_suffix}",
            description=f"RDS PostgreSQL credentials for {environment_suffix}",
            tags={
                **resource_tags,
                'Name': f"db-credentials-{environment_suffix}"
            },
            opts=ResourceOptions(parent=self)
        )

        self.db_secret_version = secretsmanager.SecretVersion(
            f"db-credentials-version-{environment_suffix}",
            secret_id=self.db_secret.id,
            secret_string=secret_value,
            opts=ResourceOptions(parent=self)
        )

        # Export values
        self.db_secret_arn = self.db_secret.arn
        self.db_username = db_username
        self.db_password_value = self.db_password.result

        self.register_outputs({
            'db_secret_arn': self.db_secret_arn,
            'db_username': pulumi.Output.secret(db_username),
            'db_password': pulumi.Output.secret(self.db_password_value)
        })

File: archive/pulumi-ts/Pr7460/lib/secrets-stack.ts
/**
 * Secrets Stack - Manages database credentials with automatic rotation.
 *
 * Features:
 * - KMS customer-managed key for encryption
 * - Secrets Manager secret for database credentials
 * - 30-day automatic rotation schedule
 */
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';
import * as random from '@pulumi/random';

export interface SecretsStackArgs {
  environmentSuffix: string;
  tags?: pulumi.Input<{ [key: string]: string }>;
}

export class SecretsStack extends pulumi.ComponentResource {
  public readonly secretArn: pulumi.Output<string>;
  public readonly kmsKeyId: pulumi.Output<string>;
  private secret: aws.secretsmanager.Secret;

  constructor(
    name: string,
    args: SecretsStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:migration:SecretsStack', name, args, opts);

    const tags = args.tags || {};

    // Create KMS key for encryption
    const kmsKey = new aws.kms.Key(
      `db-encryption-key-${args.environmentSuffix}`,
      {
        description: `KMS key for database encryption - ${args.environmentSuffix}`,
        enableKeyRotation: true,
        deletionWindowInDays: 10,
        tags: {
          ...tags,
          Name: `db-encryption-key-${args.environmentSuffix}`,
        },
      },
      { parent: this }
    );

    // Create KMS key alias
    new aws.kms.Alias(
      `db-key-alias-${args.environmentSuffix}`,
      {
        name: `alias/db-migration-${args.environmentSuffix}`,
        targetKeyId: kmsKey.keyId,
      },
      { parent: this }
    );

    // Generate secure random password using Pulumi random provider
    const dbPassword = new random.RandomPassword(
      `db-password-${args.environmentSuffix}`,
      {
        length: 32,
        special: true,
        overrideSpecial: '!#$%&*()-_=+[]{}<>:?',
      },
      { parent: this }
    );

    // Create Secrets Manager secret for database credentials
    this.secret = new aws.secretsmanager.Secret(
      `db-credentials-${args.environmentSuffix}`,
      {
        name: `db-credentials-${args.environmentSuffix}`,
        description: 'Database credentials for PostgreSQL migration',
        kmsKeyId: kmsKey.id,
        tags: {
          ...tags,
          Name: `db-credentials-${args.environmentSuffix}`,
        },
      },
      { parent: this }
    );

    // Create initial secret version with generated password
    new aws.secretsmanager.SecretVersion(
      `db-credentials-version-${args.environmentSuffix}`,
      {
        secretId: this.secret.id,
        secretString: pulumi.interpolate`{
          "username": "dbadmin",
          "password": "${dbPassword.result}",
          "engine": "postgres",
          "host": "placeholder",
          "port": 5432,
          "dbname": "migrationdb"
        }`,
      },
      { parent: this }
    );

    // Export outputs
    this.secretArn = this.secret.arn;
    // Use ARN for RDS kmsKeyId which requires an ARN format
    this.kmsKeyId = kmsKey.arn;

    this.registerOutputs({
      secretArn: this.secretArn,
      kmsKeyId: this.kmsKeyId,
    });
  }

  // Method to configure rotation after Lambda function is created
  public configureRotation(lambdaArn: pulumi.Output<string>): void {
    new aws.secretsmanager.SecretRotation(
      `db-rotation-${this.secret.name}`,
      {
        secretId: this.secret.id,
        rotationLambdaArn: lambdaArn,
        rotationRules: {
          automaticallyAfterDays: 30,
        },
      },
      { parent: this }
    );
  }
}
